File: solution/day13/src/test/java/blog/ArticleTests.java
package blog;

import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

import java.time.LocalDate;

import static blog.ArticleBuilder.anArticle;
import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

class ArticleTests {
    @Test
    void should_add_comment_in_an_article() throws CommentAlreadyExistException {
        final var article = anArticle().build();
        article.addComment(ArticleBuilder.COMMENT_TEXT, ArticleBuilder.AUTHOR);

        assertThat(article.getComments()).hasSize(1);
        assertComment(article.getComments().get(0), ArticleBuilder.COMMENT_TEXT, ArticleBuilder.AUTHOR);
    }

    @Test
    void should_add_comment_in_an_article_containing_already_a_comment() throws CommentAlreadyExistException {
        final var newComment = "Finibus Bonorum et Malorum";
        final var newAuthor = "Al Capone";

        var article = anArticle()
                .commented()
                .build();

        article.addComment(newComment, newAuthor);

        assertThat(article.getComments()).hasSize(2);
        assertComment(article.getComments().getLast(), newComment, newAuthor);
    }

    @Nested
    class Fail {
        @Test
        void when_adding_an_existing_comment() throws CommentAlreadyExistException {
            final var article = anArticle().build();
            article.addComment(ArticleBuilder.COMMENT_TEXT, ArticleBuilder.AUTHOR);

            assertThatThrownBy(() ->
                    article.addComment(ArticleBuilder.COMMENT_TEXT, ArticleBuilder.AUTHOR))
                    .isInstanceOf(CommentAlreadyExistException.class);
        }
    }

    private static void assertComment(Comment comment, String commentText, String author) {
        assertThat(comment.text()).isEqualTo(commentText);
        assertThat(comment.author()).isEqualTo(author);
        assertThat(comment.creationDate()).isBeforeOrEqualTo(LocalDate.now());
    }
}

File: solution/c#/Day13/Day13.Tests/ArticleTests.cs
using FluentAssertions;
using Xunit;
using static Day13.Tests.ArticleBuilder;

namespace Day13.Tests
{
    public class ArticleTests
    {
        [Fact]
        public void Should_Add_Comment_In_An_Article()
        {
            var article = AnArticle().Build();

            article.AddComment(CommentText, Author);

            article.Comments.Should().HaveCount(1);
            AssertComment(article.Comments[0], CommentText, Author);
        }

        [Fact]
        public void Should_Add_Comment_In_An_Article_Containing_Already_A_Comment()
        {
            const string newComment = "Finibus Bonorum et Malorum";
            const string newAuthor = "Al Capone";

            var article = AnArticle()
                .Commented()
                .Build();

            article.AddComment(newComment, newAuthor);

            article.Comments.Should().HaveCount(2);
            AssertComment(article.Comments[1], newComment, newAuthor);
        }

        private static void AssertComment(Comment comment, string expectedComment, string expectedAuthor)
        {
            comment.Text.Should().Be(expectedComment);
            comment.Author.Should().Be(expectedAuthor);
            comment.CreationDate.Should().Be(DateOnly.FromDateTime(DateTime.Now));
        }

        public class Fail
        {
            [Fact]
            public void When_Adding_An_Existing_Comment()
            {
                var article = AnArticle().Build();
                article.AddComment(CommentText, Author);

                var act = () => article.AddComment(CommentText, Author);
                act.Should().Throw<CommentAlreadyExistException>();
            }
        }
    }
}